src/manifold/deferred.clj

```diff
@@ -1386,6 +1512,152 @@
 
 (prefer-method print-method IDeferred IDeref)
 
+(defn- flatten-deferred
+  "Converts a Deferred<Deferred<T>> into a Deferred<T>"
+  [value]
+  (chain value identity))
+
+(defn- async-for
+  "Retuns a CompletionStage async version of the given operator"
+  [original]
+  (fn result
+    ([this operator]
+     (result this operator (or (ex/executor) (ex/execute-pool))))
+    ([this operator executor]
+     (flatten-deferred
+      (future-with executor (original this operator))))))
+
+(defn- async-for-dual
+  "Retuns a CompletionStage async version for the given two deferred operator"
+  [original]
+  (fn result
+    ([this other operator]
+     (result this other operator (or (ex/executor) (ex/execute-pool))))
+    ([this other operator executor]
+     (flatten-deferred
+      (future-with executor (original this other operator))))))
+
+(defmacro ^:no-doc assert-some
+  "Throws NullPointerException if any of the aruments is null."
+  [& values]
+  `(do ~@(for [value values]
+         `(when (nil? ~value)
+            (throw (NullPointerException. ~(str value " was null"))))
+         )))
```

Comment:
I assume this is to prevent NPEs from being thrown later in another thread?

You mostly have the right idea, but there's a bad practice here. You're not capturing the values before inserting them.

In this particular case, it doesn't matter, but only because you used the original form _exactly_ once. Had you inserted `value` more than once, it could be evaluated multiple times, which you rarely want. (The `str` call doesn't count, because it won't re-evaluate the argument.)

You should get into the habit of capturing, to ensure a form is evaluates only once, like:

```clojure
(defmacro ^:no-doc assert-some
  "Throws NullPointerException if any of the aruments is null."
  [& values]
  `(do
     ~@(for [value values]
         `(let [value# ~value]
            (when (nil? value#)
              (throw (NullPointerException. ~(str value " was null"))))))))
```

If I had to use `value` again, by using `value#`, I'd be reusing the already-evaluated value, instead of potentially re-evaluating.

Also, be sure trailing parens are all on the final line.

Reply:
Thank you for the advice, taking note on that. I placed these NPE assertions since the [documentation](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionStage.html) for `CompletionStage` states:

> Additionally, while arguments [...] of type T for methods accepting them may be null, passing a null value for any other parameter will result in a [NullPointerException](https://docs.oracle.com/javase/8/docs/api/java/lang/NullPointerException.html) being thrown.

And about the trailing parens, I'm checking some clojure style checking tool before writing my next line of clojure, sorry for this slip

Reply:
Well, attempting to _use_ a null will also results in an NPE being thrown, so in single-threaded scenarios, you might not care to do anything special with the arg, and just wait for it to blow up anyway. 

In multi-threaded scenarios, however, it can be harder to debug when the NPE occurs on a different thread, so checking and throwing here is fine. 

---

The best summary of the community style is at https://guide.clojure.style/. Ironically, Manifold predates a lot of that, so don't take style hints from Manifold 😄

Reply:
I took a closer look at this, and realized that you may have had it right the first time, if we tighten up what we accept.

The default macro behavior should _always_ be to capture with a `let`, since you don't know if the macro will be passed a symbol or a form, but in this case, `assert-some` is only ever used 1) by us, and 2) with fn param symbols, so we can restrict it and throw a compile-time error if the macro is passed a form instead. And in that case, we can skip the `let`, since there's no need to capture.

It would look like this:

```clojure
(defmacro ^:no-doc assert-some
  "Throws NullPointerException if any of the arguments is nil. Only works on symbols.

   Usage example:
   (defn foo [x y z]
     (assert-some x y z)
     ...)"
  [& values]
  (when (not-every? symbol? values)
    (throw (IllegalArgumentException. "assert-some only accepts symbols")))
  `(do ~@(for [value values]
           `(when (nil? value)
              (throw (NullPointerException. ~(str value " was null")))))))
```

Finally, can you move `assert-some` to the utils ns?